File: python/eo.py
#!/usr/bin/env python3
'''
The way that cumsize is computed
needs to match the way the other functions work.
In particular,
lexeo_idx_to_coord needs to recover the right parity
by adding 1 to the "fastest varying" coordinate
used in computing cumsize.

NOTE: all dimensions are assumed to be checkerboarded.
'''


def get_cumsizes(sizes):
    cumsizes = [1]
    for size in sizes:
        cumsizes.append(size * cumsizes[-1])

    return cumsizes


def lex_coord_to_idx(relative_xs, cumsizes):
    return sum(relative_x * cumsize  #
               for relative_x, cumsize  #
               in zip(relative_xs, cumsizes))


def coord_to_eo(relative_xs):
    return sum(relative_xs) % 2


def lex_coord_to_eoidx(relative_xs, cumsizes):
    idxh = lex_coord_to_idx(relative_xs, cumsizes) // 2
    eo = coord_to_eo(relative_xs)

    return eo, idxh


def lex_idx_to_coord(idx, sizes):
    coordinates = []
    i = idx
    for s in sizes:
        coordinates.append(i % s)
        i = i // s
    return coordinates


def lexeo_idx_to_coord(eo, idxh, sizes):
    coord = lex_idx_to_coord(2 * idxh, sizes)
    if eo == coord_to_eo(coord):
        return coord
    else:
        return lex_idx_to_coord(2 * idxh + 1, sizes)

File: src/geometry/eo.cpp
#include "geometry/eo.hpp"

namespace hypercubes {
namespace slow {
namespace eo {

using IntList = std::vector<int>;
IntList get_cumsizes(const IntList &sizes) {
  IntList res;
  ensuresize(res, sizes.size() + 1);
  res[0] = 1;
  for (int i = 0; i < sizes.size(); ++i)
    res[i + 1] = res[i] * sizes[i];
  return res;
}

IntList lex_idx_to_coord(int lexi, const IntList &sizes) {
  IntList coordinates;
  ensuresize(coordinates, sizes.size());
  int i = lexi;
  for (int d = 0; d < sizes.size(); ++d) {
    coordinates[d] = i % sizes[d];
    i = i / sizes[d];
  }

  return coordinates;
}

Parity coord_to_eo(const IntList &coords) {
  int sum = 0;
  for (auto x : coords)
    sum += x;
  return (sum % 2 == 0) ? Parity::EVEN : Parity::ODD;
}

int lex_coord_to_idx(const IntList &coords, const IntList &cumsizes) {
  int idx = 0;
  for (int i = 0; i < coords.size(); ++i)
    idx += coords[i] * cumsizes[i];
  return idx;
}

std::tuple<Parity, int> lex_coord_to_eoidx(const IntList &coords,
                                           const IntList &cumsizes) {
  int idxh = lex_coord_to_idx(coords, cumsizes) / 2;
  Parity p = coord_to_eo(coords);

  return std::make_tuple(p, idxh);
}

IntList lexeo_idx_to_coord(Parity eoflag, int idxh, const IntList &sizes) {
  auto coordinates = lex_idx_to_coord(2 * idxh, sizes);
  if (eoflag == coord_to_eo(coordinates))
    return coordinates;
  else
    return lex_idx_to_coord(2 * idxh + 1, sizes);
}

void ensuresize(IntList &l, int s) { l.resize(s); }

} // namespace eo
} // namespace slow
} // namespace hypercubes
